### PR TITLE
Feature: Contextual actions for vehicles grouped by shared orders

### PR DESCRIPTION
**Pull request by btzy**
## Motivation

This is a possible addition to #7028 to allow (some) contextual actions when vehicle lists are grouped by shared orders.  Contextual actions are clicks on a vehicle in the vehicle list that complete some other ongoing action, such as to clone a vehicle or to copy/share orders.

This brings the option with grouping enabled to "parity" with the non-grouped option, by supporting contextual actions wherever it is unambiguous.

## Description

The main addition of this PR is the contextual callback (`VehicleClicked()`) overload for multiple vehicles.  This PR is implemented as a generic mechanism for contextual actions on sets containing multiple vehicles, instead of just a single vehicle.  It hence does not rely on anything specific to grouping by _shared orders_, so it will still work even if in the future we have other ways to group vehicles.

I've implemented this whole feature outside of the `DoCommand` system.  I'm not sure if this is the best way, but since vehicle grouping is a purely UI feature (kind of a "view" over the real list of vehicles), it might be arguable that this resolution of contextual actions should also happen outside of the command system too.  Also I'm not sure if there's a simpler way to implement this.

## How to test

1. Load a game, open the vehicle list of some station with many routes passing through it, and set the vehicle list to group by shared orders.
2. Create a new vehicle, open the order list, click "Go to", then click any row in the vehicle list from step 1.  Observe that the order list is copied to the new vehicle.
3. Same as step 2, but hold down CTRL when clicking the row in the vehicle list.  Observe that the new vehicle is now sharing orders.
4. Create a depot, click "Clone Vehicle", then click any row in the vehicle list from step 1.  If all vehicles in the vehicle list have the same sequence of engines, then the vehicle will be cloned.  Otherwise an error will be displayed.
5. Same as step 4, but hold down CTRL when clicking the row in the vehicle list.  You should observe the same thing as step 4, but the new vehicle is now sharing orders.

Note:  I'm not sure if checking that two vehicles have the same sequence of engines is the correct way to check if two vehicles are "equal" (for steps 4 and 5).  Please correct me if it isn't!

### CODE REVIEW

**Comment by 2TallTyler** [comment]
I often group vehicles by shared orders and would find this useful. If you rebase and fix it (the preview build failed), I promise to review it. 😄

**Comment by 2TallTyler** [comment]
I wasn't a developer two years ago. 🙂 But now I am and I'm trying to review the low-hanging fruit PRs that are within my abilities.

**Comment by btzy** [comment]
@2TallTyler Done rebasing.

**Comment by 2TallTyler** [comment]
There were some conflicts and the CI checks didn't run so I rebased. I'll try to review tomorrow.

File: src/depot_gui.cpp [hunk 1]
```diff
@@ -916,6 +918,52 @@ struct DepotWindow : Window {
...
+	 * @return Returns true if clone/copy succeeded, false otherwise.
+	 */
+	bool OnVehicleSelect(VehicleList::const_iterator begin, VehicleList::const_iterator end) override
+	{
+		if (!_ctrl_pressed) {
+			/* If CTRL is not pressed: If all the vehicles in this list have the same orders, then copy orders */
+			if (AllEqual(begin, end, [](const Vehicle *v1, const Vehicle *v2) {
+				return VehiclesHaveSameEngineList(v1, v2);
+			})) {
+				if (AllEqual(begin, end, [](const Vehicle *v1, const Vehicle *v2) {
```
Comment (2TallTyler) [inline src/depot_gui.cpp:934]: Maybe I'm reading this wrong, but is it possible to fulfill this condition? Wouldn't it have returned on line 930?

File: src/depot_gui.cpp [hunk 2]
```diff
@@ -916,6 +918,52 @@ struct DepotWindow : Window {
 		return true;
 	}
 
+	/**
+	 * Clones a vehicle from a vehicle list.
```
Comment (2TallTyler) [inline src/depot_gui.cpp]: Because this is an override the function name can't be very specific. Can you specify in this doxygen header when it fires? E.g. "When the X button is clicked on and Y is true, try to clone a vehicle from a vehicle list." As it is I'm confused. 🙂

File: src/depot_gui.cpp [hunk 3]
```diff
@@ -916,6 +918,52 @@ struct DepotWindow : Window {
...
+			/* If CTRL is not pressed: If all the vehicles in this list have the same orders, then copy orders */
+			if (AllEqual(begin, end, [](const Vehicle *v1, const Vehicle *v2) {
+				return VehiclesHaveSameEngineList(v1, v2);
+			})) {
+				if (AllEqual(begin, end, [](const Vehicle *v1, const Vehicle *v2) {
+					return VehiclesHaveSameOrderList(v1, v2);
+				})) {
+					OnVehicleSelect(*begin);
+				}
+				else {
```
Comment (2TallTyler) [inline src/depot_gui.cpp]: Wrong format, should be `} else {` all on one line. This applies several places in this PR. 🙂

**Comment by btzy** [comment]
Sorry, the earlier rebase didn't update it much (I mistakenly rebased on origin/master instead of upstream/master).  I've rebased it properly this time.

File: src/depot_gui.cpp [hunk 4]
```diff
@@ -916,6 +918,52 @@ struct DepotWindow : Window {
...
+					ShowErrorMessage(STR_ERROR_CAN_T_BUY_TRAIN + (*begin)->type, STR_ERROR_CAN_T_COPY_ORDER_VEHICLE_LIST, WL_INFO);
+				}
+			}
+			else {
+				ShowErrorMessage(STR_ERROR_CAN_T_BUY_TRAIN + (*begin)->type, STR_ERROR_CAN_T_CLONE_VEHICLE_LIST, WL_INFO);
```
Comment (2TallTyler) [inline src/depot_gui.cpp:942]: Shouldn't this (and the other failed paths) `return false` after showing the error message, instead of falling through to the `return true` at the bottom?

File: src/order_gui.cpp [hunk 5]
```diff
@@ -1465,6 +1468,41 @@ struct OrdersWindow : public Window {
 		return true;
 	}
 
+	/**
+	 * Clones an order list from a vehicle list.
```
Comment (2TallTyler) [inline src/order_gui.cpp]: This could also use a description of when this fires, as above.

File: src/vehicle_gui.cpp [hunk 6]
```diff
@@ -3250,6 +3250,31 @@ bool VehicleClicked(const Vehicle *v)
 	return _thd.GetCallbackWnd()->OnVehicleSelect(v);
 }
 
+/**
+ * Dispatch a "vehicle group selected" event if any window waits for it.
+ * @param begin iterator to the start of the range of vehicles
+ * @param end iterator to the end of the range of vehicles
+ * @return did any window accept vehicle group selection?
```
Comment (2TallTyler) [inline src/vehicle_gui.cpp:3280]: Should be `@return True if X, else false` or similar, I think

File: src/vehicle_gui.cpp [hunk 7]
```diff
@@ -3250,6 +3250,31 @@ bool VehicleClicked(const Vehicle *v)
...
+ */
+bool VehicleClicked(VehicleList::const_iterator begin, VehicleList::const_iterator end)
+{
+	assert(begin != end);
+	if (!(_thd.place_mode & HT_VEHICLE)) return false;
```
Comment (2TallTyler) [inline src/vehicle_gui.cpp:3285]: Probably needs a comment explaining what this means

File: src/vehicle_gui.cpp [hunk 8]
```diff
@@ -3250,6 +3250,31 @@ bool VehicleClicked(const Vehicle *v)
...
+
+	/* If there is only one vehicle in the group, act as if we clicked a single vehicle */
+	if (begin + 1 == end) return _thd.GetCallbackWnd()->OnVehicleSelect(*begin);
+
+	return _thd.GetCallbackWnd()->OnVehicleSelect(begin, end);
+}
+
+/**
+ * Helper function that converts a GUIVehicleGroup to a vehicle range.
```
Comment (2TallTyler) [inline src/vehicle_gui.cpp]: Needs to list params

File: src/window_gui.h [hunk 9]
```diff
@@ -755,6 +758,13 @@ struct Window : ZeroedMemoryAllocator {
 	 */
 	virtual bool OnVehicleSelect(const struct Vehicle *v) { return false; }
 
+	/**
+	 * The user clicked on a vehicle while HT_VEHICLE has been set.
+	 * @param v clicked vehicle. It is guaranteed to be v->IsPrimaryVehicle() == true
```
Comment (2TallTyler) [inline src/window_gui.h]: Doxygen headers should be written in plain English, with no code included.

File: src/window_gui.h [hunk 10]
```diff
@@ -11,7 +11,10 @@
 #define WINDOW_GUI_H
 
 #include <list>
+#include <algorithm>
```
Comment (2TallTyler) [inline src/window_gui.h:14]: Not a critique, but a question: What are these? I've never seen references to these before in OpenTTD. (Remember I'm new to the development team. 🙂 )

File: src/depot_gui.cpp (hunk 1 above)
Comment (btzy) [reply src/depot_gui.cpp:934]: In English, lines 929 to 943 are sort of equivalent to this:
```
if all vehicles have the same engine list:
    if all vehicles have the same order list:
        OnVehicleSelect(...);
    else:
        ShowErrorMessage(..., STR_ERROR_CAN_T_COPY_ORDER_VEHICLE_LIST, ...);
else:
    ShowErrorMessage(..., STR_ERROR_CAN_T_CLONE_ORDER_VEHICLE_LIST, ...);
```
So the code is correct.

File: src/depot_gui.cpp (hunk 3 above)
Comment (btzy) [reply src/depot_gui.cpp]: Yeah, Visual Studio seems to always do this.  I'll fix it.

File: src/depot_gui.cpp (hunk 4 above)
Comment (btzy) [reply src/depot_gui.cpp:942]: Nope that would be wrong.  OnVehicleSelect() returns true to tell the caller that the mouse click is handled by the active contextual action, and so the click shouldn't be re-handled by the default action.  In this case, the active contextual action is to clone/copy a vehicle (because the "Clone Vehicle" button was pressed), and we shouldn't do the default action (e.g. open the vehicle window) even if the contextual action fails to be performed.  Showing an error message (and not doing anything else) is the correct and expected behaviour here.

In the part of my PR in group_gui.cpp and vehicle_gui.cpp you will see some code that looks like this:
```
if (!VehicleClicked(vehgroup)) {
    // <-- do the default action here (i.e. open either the vehicle view window or vehicle list window)
}
```
The above code does the default action if the active contextual action did not handle the mouse click.

File: src/depot_gui.cpp (hunk 2 above)
Comment (btzy) [reply src/depot_gui.cpp]: Done.

File: src/order_gui.cpp (hunk 5 above)
Comment (btzy) [reply src/order_gui.cpp]: Done.

File: src/vehicle_gui.cpp (hunk 6 above)
Comment (btzy) [reply src/vehicle_gui.cpp:3280]: I'm just copying the doc for the `bool VehicleClicked(const Vehicle *v)` overload just above this code.  We should either change none or both, and I went with none because I don't think we should make unrelated changes to the PR.

File: src/vehicle_gui.cpp (hunk 7 above)
Comment (btzy) [reply src/vehicle_gui.cpp:3285]: Same as above, this is copied code.  I have no clue what it means right now, but if you really want an answer I could try reading through more code to figure out an answer.

File: src/vehicle_gui.cpp (hunk 8 above)
Comment (btzy) [reply src/vehicle_gui.cpp]: Done

File: src/window_gui.h (hunk 10 above)
Comment (btzy) [reply src/window_gui.h:14]: This is a standard C++ header, which comes with all standards-compliant C++ compilers.  See [C++ reference page for `<algorithm>`](https://en.cppreference.com/w/cpp/header/algorithm).

In this case `<algorithm>` is needed for `std::adjacent_find` and `<functional>` is needed for `std::not_fn`.

File: src/window_gui.h (hunk 9 above)
Comment (btzy) [reply src/window_gui.h]: This is copied from the `bool OnVehicleSelect(const struct Vehicle *v);` function immediately above.

But anyway what we actually want is a precondition, so I've changed it and the function immediately above.

**Comment by btzy** [comment]
Thanks for the review!

File: src/depot_gui.cpp (hunk 1 above)
Comment (glx22) [reply src/depot_gui.cpp:934]: Line 930 is the body of the lambda passed to `AllEqual`

Comment (2TallTyler) [reply src/depot_gui.cpp:934]: Ah, thanks @glx22 for the clarification.

File: src/window_gui.h (hunk 10 above)
Comment (2TallTyler) [reply src/window_gui.h:14]: Thanks for the explanation.

**Review by 2TallTyler — APPROVED** [review]
Works as described. Thanks for the detailed test instructions!

Code looks fine to me, but should definitely get a second pass before merging.
